llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -4242,8 +4242,19 @@ SelectionDAG::computeOverflowForSignedAdd(SDValue N0, SDValue N1) const {
   if (ComputeNumSignBits(N0) > 1 && ComputeNumSignBits(N1) > 1)
     return OFK_Never;
 
-  // TODO: Add ConstantRange::signedAddMayOverflow handling.
-  return OFK_Sometime;
+  // smulhi + any value never overflow
```

Comment:
This is clearly nonsense. smulhi plus _some_ values can overflow. The useful thing you know about smulhi is that it is in a restricted range, e.g. (for 32 bits) -0x40000000 to +0x3fffffff.

Reply:
If 0x3fffffff is the top value, then how is 0x40000000 not valid signed in 32 bits?

Reply:
> The useful thing you know about smulhi is that it is in a restricted range, e.g. (for 32 bits) -0x40000000 to +0x3fffffff.

Isn't the maximum positive value 0x40000000 from 0x8000000*0x8000000?

Reply:
0x80000000 is a negative number

Reply:
You've been asked multiple times to stop resolving conversations with your reply.

0x8000000 is a negative number, but two negatives multiplied together is a positive number. 0x8000000*0x80000000 is 0x4000000000000000 a positive number.